plugins/pulp_rpm/plugins/importers/yum/purge.py

```diff
@@ -400,55 +401,39 @@ def _duplicate_key_id_generator_aggregation(unit, fields):
     # It is a happy coincidence that the two dicts are the same values
     pipeline_opts = dict(zip(fields, repeat(1)))
 
-    # for $sort, this indicates ascending sort for nevra fields
-    sort = {'$sort': pipeline_opts}
+    lookup = {'$lookup': {'from': unit._meta['collection'],
+                          'localField': 'unit_id', 'foreignField': '_id', 'as': 'units'}}
+
+    match = {'$match': {'repo_id': repo_id, 'unit_type_id': unit._content_type_id.default}}
+
+    unwind = {'$unwind': '$units'}
+
+    replace_root = {'$replaceRoot': {'newRoot': '$units'}}
```

Comment:
This operation was introduced in MongoDB 3.4  https://docs.mongodb.com/manual/reference/operator/aggregation/replaceRoot/

Currently, pulp indirectly requires MongDB 2.4+ by requiring pymongo >= 3.0.0
https://github.com/pulp/pulp-packaging/blob/master/packages/pulp/pulp.spec#L386

MongoDB 2.4 support got dropped in pymongo 3.3. https://pymongo.readthedocs.io/en/stable/changelog.html#changes-in-version-3-3

If we need to require MongoDB 3.4, we need to do it explicitly, since pymongo 3.x supports MongoDB 2.6+. And also drop the unused code paths (like the map reduce approach).

My personal opinion is that it might be a too large of a change for a Z release, and/or releasing Y stream for the project in maintenance mode.
Any thoughts?

Reply:
I share the concern about how large the change is.  It seems risky this late into the lifecycle.

I'm not that concerned about dropping support for old MongoDB.  Anything older than MongoDB 3.6 is unsupported upstream (for more than a year, 3.4 support ended Jan 2020), and all the SCLs older than 3.2 are marked EOL.  As far as migrating to Pulp 3, it might be good to have a higher baseline for the supported versions anyway.

But I do think it would require a 2.22.  Maybe we need to do that anyway: https://github.com/pulp/pulp/pull/4018

Reply:
@goosemania Maybe we need to be careful about the versions of libraries we use in the migration plugin if we want to be able to migrate from these old databases.

Reply:
Changed code to avoid using newRoot.